language-support/hs/bindings/examples/nim-console/src/Local.hs

```diff
@@ -0,0 +1,191 @@
+import Data.Map.Strict (Map)
+
+import Domain
+import External
+
+----------------------------------------------------------------------
+-- local command, map to X in context of state
+
+data LCommand
+    = OfferNewGameToAnyone
+    | OfferGameL Player
+    | AcceptOfferL Onum
+    | MakeMoveL Gnum Move
+    deriving Show
+
+externCommand :: Player -> State -> LCommand -> Maybe XCommand
+externCommand who state@State{knownPlayers} = \case
+    OfferNewGameToAnyone -> do
+        return $ OfferGame (Offer {from = who, to = knownPlayers \\ [who] })
+    OfferGameL player -> do
+        return $ OfferGame (Offer {from = who, to = [player] }) --can play self!
+    AcceptOfferL lid -> do
+        xid <- externOid state lid
+        return $ AcceptOffer who xid
+    MakeMoveL lid move -> do
+        xid <- externGid state lid
+        return $ MakeMove who xid move
+
+externOid :: State -> Onum -> Maybe Xoid
+externOid State{offers} low = (fmap fst . List.find (\(_,(l,_,status)) -> l==low && isOpen status) . Map.toList) offers
+
+externGid :: State -> Gnum -> Maybe Xgid
+externGid State{games} low = (fmap fst . List.find (\(_,(l,_,status)) -> l==low && isOpen status) . Map.toList) games
+
+----------------------------------------------------------------------
+-- local trans, for reporting what happing in terms of local oid/gid
+data LTrans
+    = NewOfferL Onum Offer
+    | OfferNowUnavailable Onum Offer
+    | NewGameL Gnum Game
+    | GameMoveL Gnum Game
+    deriving Show
+
+-- TODO: return list, then caller can make random choice!
+lookForAnAction :: State -> Maybe LCommand
+lookForAnAction State{whoami,offers,games} =
+    listToMaybe $
+        -- prefer to play a game move..
+        mapMaybe
+        (\(onum,offer,status) ->
+                if whoami `elem` to offer && isOpen status
+                then Just $ AcceptOfferL onum
+                else Nothing
+        ) (Map.elems offers)
+        ++
+        -- otherwise accept any pending offer
+        List.concatMap
+        -- randomize move order here !
+        (\(gnum,game,status) ->
+                if whoami == p1 game && isOpen status
+                then map (MakeMoveL gnum) (legalMovesOfGame game)
+                else []
+        ) (Map.elems games)
+
+----------------------------------------------------------------------
+-- local state, accumulates external transitions
+
+data State = State {
+    whoami :: Player,
+    knownPlayers :: [Player],
+    offers :: Map Xoid (Onum,Offer,Status),
+    games :: Map Xgid (Gnum,Game,Status),
+
+    -- TODO: share the next number thing for offers and games
+    -- when an offer is converted to a game, keep the number
+    -- this mean a X-NewGame wil need to refernce the offer is comes from
+    nextOfferId :: Onum,
+    nextGameId :: Gnum
+    }
+    deriving (Show)
+
+data Status = Open | Closed
+    deriving (Show)
+
+isOpen :: Status -> Bool
+isOpen = \case Open -> True; Closed -> False
```

Comment:
Is this function useful? Why not deriving Eq and then do `== Open` when you need this.

Reply:
ok